zigpy/application/state.py

```diff
@@ -0,0 +1,220 @@
+"""Classes to implement status of the application controller."""
+
+from dataclasses import InitVar, dataclass, field
+import functools
+from typing import Any, Dict, Iterable, List, Optional
+
+import zigpy.types as t
+import zigpy.zdo.types as zdo_t
+
+
+@dataclass
+class NodeInfo:
+    """Controller Application network Node information."""
+
+    nwk: t.NWK = t.NWK(0xFFFE)
+    ieee: Optional[t.EUI64] = None
+    logical_type: Optional[zdo_t.LogicalType] = None
+
+    def __post_init__(self) -> None:
+        """Initialize instance."""
+        if self.ieee is None:
+            self.ieee = t.EUI64.convert("ff:ff:ff:ff:ff:ff:ff:ff")
+        if self.logical_type is None:
+            self.logical_type = zdo_t.LogicalType.Reserved7
+
+
+@dataclass
+class NetworkInformation:
+    """Network information."""
+
+    extended_pan_id: Optional[t.ExtendedPanId] = None
+    pan_id: Optional[t.PanId] = 0xFFFE
+    nwk_update_id: Optional[t.uint8_t] = 0x00
+    nwk_manager_id: Optional[t.NWK] = t.NWK(0xFFFE)
```

Comment:
Is this an EZSP-specific property? Or for distributed networks?

Reply:
This is Network Manager and is Zigbee specific. Part of the persistent data, each node holds

Reply:
I think this change is useful. What do you think about adding a few more fields?

```Python
@dataclass
class NodeInfo:
    """Controller Application network Node information."""

    nwk: t.NWK | None
    ieee: t.EUI64
    logical_type: zdo_t.LogicalType

    # New
    aps_link_key: t.KeyData | None
    aps_tx_counter: t.uint32_t | None
    aps_rx_counter: t.uint32_t | None


@dataclass
class NetworkInformation:
    """Network information."""

    extended_pan_id: t.ExtendedPanId
    pan_id: t.PanId
    nwk_update_id: t.uint8_t
    nwk_manager_id: t.NWK
    channel: t.uint8_t
    channel_mask: t.Channels

    # New
    security_level: t.uint8_t
    nwk_key: t.KeyData
    nwk_key_seq: t.uint8_t
    nwk_key_frame_counter: t.uint32_t

    associated_nodes: list[NodeInfo]
```

If radio libraries formed networks from a `NetworkInformation` object and returned one when starting up an adapter, this would allow for seamless backup/restore between coordinators and for network settings to easily be changed.

Reply:
That's fine by me. Originally this was intended to be passed to UI to display information about the controller, that's why sensitive info was excluded, but we can filter that out and have a single object for both purposes.

Reply:
Do you want me to make a separate PR with just the network info changes or do you want to merge one of your counter branches instead?

Reply:
I need to create a new PR. probably over the weekend. I'll incorporate your changes

Reply:
```
    # New
    aps_link_key: t.KeyData | None
    aps_tx_counter: t.uint32_t | None
    aps_rx_counter: t.uint32_t | None
```

Should this be a Trust Center Link Key instead? And maybe add a list or dict of APS link keys, because usually APS key is unique per specific destination? i.e.  if you a device on the networks wants to use encrypted APS with another device (not Coordinator), then it is going to use the APS key specific to that destination device only?